### PR TITLE
feat: Verify guest links are allowed for foreign conversation SQSERVICES-1340

### PR DESCRIPTION
**Pull request by agisilaos**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1340" title="SQSERVICES-1340" target="_blank">TaskSQSERVICES-1340</a>  [iOS] Verify guest links are allowed for "foreign" conversation
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

With this PR we support a new endpoint in SE regarding the status for the guest link. So to do that we need to do a couple of things:

1. Create a new request `guestLinkStatusRequest`  for the new endpoint.
2. Create a new public method called `checkGuestLinkStatus` and return the appropriate result based on the `httpStatus`, the `payload` etc..
3. Add a new error response based on the information we have on Swagger called `no-conversation`

### Notes

- Added tests for the new request and the new error response
- [ ] Need to add Integration Tests as soon as I make the changes in Mock Transport
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: Source/Data Model/Conversation+AccessMode.swift [hunk 1]
```diff
@@ -135,6 +138,33 @@ extension ZMConversation {
         userSession.transportSession.enqueueOneTime(request)
     }
 
+    /// Checks the status of the guest link. If it's either enabled or disabled
+    public func checkGuestLinkStatus(in userSession: ZMUserSession, _ completion: @escaping (Result<Bool>) -> Void) {
```
Comment (johnxnguyen) [inline Source/Data Model/Conversation+AccessMode.swift:142]: Suggestion: 

```swift
public func canGenerateGuestLink(in session: ZMUserSession, completion: @escaping (Result<Bool>)  -> Void)
````

I think it reads better and is more obvious what the method does. With `checkGuestLinkStatus`, it's not clear what "status" means without further context. For instance, does it refer to an existing guest link? Does it refer to whether the feature is available?

File: Source/Data Model/Conversation+AccessMode.swift [hunk 2]
```diff
@@ -135,6 +138,33 @@ extension ZMConversation {
...
+
+        let request = WirelessRequestFactory.guestLinkStatusRequest(for: self)
```
Comment (johnxnguyen) [inline Source/Data Model/Conversation+AccessMode.swift:144]: Suggestion:

```swift
        let request = WirelessRequestFactory.guestLinkFeatureStatusRequest(for: self)
```

File: Source/Data Model/Conversation+AccessMode.swift [hunk 3]
```diff
@@ -135,6 +138,33 @@ extension ZMConversation {
...
+
+        request.add(ZMCompletionHandler(on: managedObjectContext!) { response in
+            switch response.httpStatus {
+            case 200:
+                if let payload = response.payload?.asDictionary(),
+                   let data = payload["status"] as? String {
+                    completion(.success(data == "enabled"))
+                } else {
+                    completion(.failure(WirelessLinkError.invalidResponse))
+                }
```
Comment (johnxnguyen) [inline Source/Data Model/Conversation+AccessMode.swift:154]: I think a guard statement is more readable here:

```swift
guard
    let payload = response.payload?.asDictionary(),
    let data = payload["status"] as? String
else {
    completion(.failure(WirelessLinkError.invalidResponse))
}
```

File: Source/Data Model/Conversation+AccessMode.swift [hunk 4]
```diff
@@ -135,6 +138,34 @@ extension ZMConversation {
...
+            switch response.httpStatus {
+            case 200:
+                guard let payload = response.payload?.asDictionary(),
+                      let data = payload["status"] as? String
+                else {
+                    return completion(.failure(WirelessLinkError.invalidResponse))
+                }
+                return completion(.success(data == "enabled"))
+                
+            case 404:
```
Comment (johnxnguyen) [inline Source/Data Model/Conversation+AccessMode.swift:156]:
```suggestion
                guard 
                    let payload = response.payload?.asDictionary(),
                    let data = payload["status"] as? String
                else {
                    return completion(.failure(WirelessLinkError.invalidResponse))
                }

                return completion(.success(data == "enabled"))

            case 404:
```

**Review by johnxnguyen — APPROVED** [review]
You still have lint issues.

**Comment by agisilaos** [comment]
release major

**Comment by johnxnguyen** [comment]
release major

**Pull request merged**